nexus/src/db/datastore/silo_user.rs

```diff
@@ -202,6 +212,96 @@ impl DataStore {
+        db_silo_user: &SiloUser,
+        db_silo_user_password_hash: Option<SiloUserPasswordHash>,
+    ) -> UpdateResult<()> {
+        opctx.authorize(authz::Action::Modify, authz_silo_user).await?;
+
+        // Verify that the various objects we're given actually match up.
+        // Failures here reflect bugs, not bad input.
+        bail_unless!(db_silo.id() == db_silo_user.silo_id);
+        bail_unless!(db_silo_user.id() == authz_silo_user.id());
+        if let Some(db_silo_user_password_hash) = &db_silo_user_password_hash {
+            bail_unless!(
+                db_silo_user_password_hash.silo_user_id == db_silo_user.id()
+            );
+        }
+
+        // Verify that this Silo supports setting local passwords on users.
+        // The caller is supposed to have verified this already.
+        bail_unless!(db_silo.user_provision_type == UserProvisionType::ApiOnly);
+
+        use db::schema::silo_user_password_hash::dsl;
+
+        if let Some(db_silo_user_password_hash) = db_silo_user_password_hash {
+            let hash_for_update = db_silo_user_password_hash.hash.clone();
+            diesel::insert_into(dsl::silo_user_password_hash)
+                .values(db_silo_user_password_hash)
+                .on_conflict(dsl::silo_user_id)
+                .do_update()
+                .set(SiloUserPasswordUpdate::new(hash_for_update))
+                .execute_async(self.pool_authorized(opctx).await?)
+                .await
+                .map_err(|e| {
+                    public_error_from_diesel_pool(e, ErrorHandler::Server)
+                })?;
+        } else {
+            diesel::delete(dsl::silo_user_password_hash)
+                .filter(dsl::silo_user_id.eq(authz_silo_user.id()))
+                .execute_async(self.pool_authorized(opctx).await?)
+                .await
+                .map_err(|e| {
+                    public_error_from_diesel_pool(e, ErrorHandler::Server)
+                })?;
+        }
+
+        Ok(())
+    }
+
+    /// Fetches the stored password hash for a given Silo user
+    ///
+    /// Returns:
+    ///
+    /// - `Ok(Some(SiloUserPasswordHash))` if the hash was found
+    /// - `Ok(None)` if we successfully queried for it but found none
+    /// - `Err(...)` otherwise
+    pub async fn silo_user_password_hash_fetch(
+        &self,
+        opctx: &OpContext,
+        authz_silo_user: &authz::SiloUser,
+    ) -> LookupResult<Option<SiloUserPasswordHash>> {
+        // Although it should be safe enough to leak a password hash, there's no
+        // reason to give it out to people who only have "read" access to the
+        // user.  So we check for "modify" here.  (Arguably we could consider
+        // the password a separate resource nested under the user, so that this
+        // would be a check on "read" of the user's password.  This doesn't seem
+        // worth the effort right now.)
+        opctx.authorize(authz::Action::Modify, authz_silo_user).await?;
```

Comment:
If this `.authorize` call fails, e.g., because the user only has `read` permission for whatever reason, then the caller (`silo_user_password_verify`) will not attempt to verify or create a password hash. Are there timing attacks that could possibly be mounted because of this?

Reply:
This is a good question.  Today, the only code path that gets here is the `POST /login/{silo_name}/local` API endpoint, which uses the built-in "external authenticator" user's OpContext.  So today, this authz check will never fail.  It's here to protect us if someone were to (accidentally?) start calling this function from some other code path.  We still don't want to expose a timing attack if that happens.  But it's hard to infer what information someone could learn in that case.  I _think_ all you can learn here is that you're not allowed to verify a password for that user (which is the same thing you learn from the error code).